test-framework/test-suites/integration/tests/add/test_add_host.py

```diff
@@ -0,0 +1,51 @@
+import pytest
+
+@pytest.mark.usefixtures("revert_database")
+class TestAddHost:
+
+	HOST_POSITIVE_TEST_DATA = [
+		('backend-0-0','','add_host_with_standard_naming_convention.json',),
+		('backend_4_3','appliance=backend rack=4 rank=3','add_host_with_appliance_rack_rank.json',),
+		('backend-7-7','appliance=backend rack=5 rank=5','add_host_with_standard_naming_and_appliance_rack_rank.json',)
+		]
+
+	@pytest.mark.parametrize("host_name, add_params, output_file", HOST_POSITIVE_TEST_DATA)
+	def test_add_host_positive_behaviour(self, host, host_name, add_params, output_file):
+		dirn = '/export/test-files/add/'
+		expected_output = open(dirn + output_file).read()
+
+		result = host.run(f'stack add host {host_name} {add_params}')
+		assert result.rc == 0
+
+		result = host.run(f'stack list host {host_name} output-format=json')
+		assert result.rc == 0
+
+		assert self.validate_add_host_positive_behaviour_json(result.stdout, expected_output)
+
+
+	
+
+	HOST_NEGATIVE_TEST_DATA = [
+		('','','',),
+		('backend','','',),
+		('backend', 'appliance=backend','',),
+		('backend', 'appliance=backend rack=4','',),
+		('backend', 'appliance=backend rank=4','',)
+		]
+
+	@pytest.mark.parametrize("host_name, add_params, output_file", HOST_NEGATIVE_TEST_DATA)
+	def test_add_host_negative_behaviour(self, host, host_name, add_params, output_file):
+		result = host.run(f'stack add host {host_name} {add_params}')
+		assert result.rc != 0
+
+
+
+
+
+	def validate_add_host_positive_behaviour_json(self, stdout_output, expected_output):
+		# Remove OS field from Stdout
+		start_index = stdout_output.find("\"os\": ")+7
```

Comment:
I see what you're doing here: the OS field will vary depending on which OS the test is running against, meaning your `expected_output` will always be wrong on one OS.  Since you're not testing for OS here, that's reasonable.

Perhaps instead of this, since they're both actually JSON you should `json.loads()` on both of them and then delete the `os` key there, then compare.  It's probably less efficient, but it's also less brittle and more readable.

Reply:
@caladd, thoughts?

Reply:
For the wsclient I just added, I got the OS from testinfa. Since our OS is either going to be `redhat` or `sles` it was clean:
```
		if host.system_info.distribution == "centos":
			os = "redhat"
		else:
			os = "sles"
		
		result = host.run("wsclient list host")
		assert result.rc == 0
		assert json.loads(result.stdout) == [
			{
				"host": "frontend-0-0",
				"rack": "0",
				"rank": "0",
				"appliance": "frontend",
				"os": os,
				"box": "default",
				"environment": None,
				"osaction": "default",
				"installaction": "default",
				"status": "deprecated",
				"comment": None
			}
		]
```

Reply:
The only thing I don't like about that is that it isn't the same way that Stacki determines (rightly or wrongly) what the OS is.  I'm not sure that alone is reason to avoid it, however.

It's also another place where we've hardcoded `centos==redhat`.

Reply:
Agreed. If the tests are ran under non-CentOS flavored Redhat then the test will fail. However, I feel like that might be a case of "we'll cross that bridge when we get there".

Since the OS really isn't important, your suggested approach of doing a `json.loads` on both then del the `os` entry would be nice and clean. It has the added benefit of confirming `output-format=json` is actually producing valid JSON